Entities/Activity.cpp

```diff
@@ -286,9 +286,12 @@ void Activity::Clear() {
 
 		// Intentionally doing all players, all need controllers
 		for (short player = Players::PlayerOne; player < Players::MaxPlayerCount; ++player) {
+			const short playerScreen = ScreenOfPlayer(player);
```

Comment:
I don't quite understand what you did here.

Reply:
When player is a cpu, then ScreenOfPlayer returns -1.
SetScreenOcclusion() tries to access an array element at that index. That's undefined behavior.
I ensured that doesn't happen.
The variable definition is just for convenience.

Reply:
As I've mentioned before, there's not much reason to bother consting playerScreen here. No harm done though really.

Reply:
`player` in this loop cannot be -1 (CPU player), so `ScreenOfPlayer` can't return -1 and `SetScreenOcclusion` can never access index -1 because of that.
So I don't see how this helps anything being done here, plus it's still being done the "unsafe" way in other classes.
I can see this being an issue if -1 is passed in manually or with a loop starting at -1, so the better fix would be adding protection to `SetScreenOcclusion`.

Reply:
Any player can be an AI and if one of the 0-3 players is an AI then `ScreenOfPlayer` returns `m_PlayerScreen[player]`.
That variable is defined:
```short m_PlayerScreen[Players::MaxPlayerCount]; //!< The screen index of each player - only applicable to human players. -1 if AI or other.```
This is why `SetScreenOcclusion` can get an input of -1 in this specific case.
_Side note: when I tested this while working on the pr, I got a compilation error before the fix. Now when I test, it compiles but crashes when you try to play with an AI as player 0-3._
I believe the protection should be outside `SetScreenOcclusion` (working on assumption that when we call it we know what we're doing).
Did I convince you, or you still want the protection inside the function?

Reply:
Maxim you're mixing up screen and player here. Player can only be 0 - MaxPlayer, but AI players have no screen so their screen returned will be -1.

As for where the safety check should be, I think it's okay to leave it outside of SetScreenOcclusion. If someone screws that up they'll probably want to know about it and fix it, and silently ignoring that screwup might make that difficult.
What's up with the crash though, that sounds bad? Or are you saying that's in development??

Reply:
It's not about convincing me, it's about pinpointing the root of the problem. Plus my *DanCode*:tm: senses are tingling.

My bad on the CPU player number, it's not `-1` it's actually `4`. `-1` is disabled player.
The AI cannot be `Player` 0-3 because they are reserved for human players only, it is always player 4 but it can be on any `Team`.
Player 4 is always active, even in `Test Activity` or human-only splitscreen where there is no AI.
In combat benchmark, if you check for active players you will get this:
![ScreenDump000](https://user-images.githubusercontent.com/50018358/101211303-a90ed480-367f-11eb-8b2b-8127a079fa42.png)
So you can see in fact, that while the AI occupies *teams* 1 and 2, it is still *player* 4.

So technically you're right about what is happening but not why it's happening.
Why it's happening is because players 1-3 are disabled, the value in `m_PlayerScreen[]` at those indexes is -1 because this array does not resize to the number of active human players (it probably should), so then, the default `Vector` copy-assignment operator fails to assign to index -1 in `m_ScreenOcclusion[]`.

## BUT,
This was never an issue before somehow, and reverting the copy-assignment operator back to:
```c++
	Vector & operator=(const Vector &rhs) {
		if (*this != rhs) {
			m_X = rhs.m_X;
			m_Y = rhs.m_Y;
		}
		return *this;
	}
```
fixes the crash.
Fuck if I know how Dan made it work, but assigning to index -1 just does it's thing and does nothing without crashing. :shrug: 

So anyway, now that we both understand what's happening I am not "convinced". ¯\\\_(ツ)_/¯
For now, please revert to the old copy-constructor definition to avoid this potentially popping up from some other place Dan neglected.

I'll open an issue about using a non-negative index for a garbage screen/disabled player once I figure out how it should be done.

Reply:
Yeah, you're correct that AI takes the spot of MaxPlayerCount (currently 4) and NoPlayer is -1. Either way, for CPU players, screen -1 is returned cause the check inside GetScreenOfPlayer stops before MaxPlayerCount.

As for the crash, I think what's going on is:
1. SetScreenOcclusion does m_ScreenOcclusion[screen] rather than m_ScreenOcclusion.at(screen) or something. Using the square bracket operator with invalid indices is undefined behaviour (see [here](https://stackoverflow.com/questions/6911289/should-i-use-stdvectorat-in-my-code)).

2. Screen occlusion is now set to a zero Vector. In the old version there's a check at the beginning of the `=` operator to only do sutff if `*this != rhs`, which in turn returns `!(lhs == rhs)`, which in turn returns `lhs.m_X == rhs.m_X && lhs.m_Y == rhs.m_Y`. Now this is where I'm a little less clear on things, but I think the undefined behaviour of the [] operator is basically giving us some junk reference Vector to compare with (or maybe making one for us on demand), that has 0 as its X and Y value. And since we're assigning a zero Vector, this junk Vector is considered equal to it, so no actual assignment attempts are happening and we're not getting a crash.

I would say it's worth making this change to the assignment operator so we can find these types of crashes and fix them, but it would be better to make that a separate issue so this doesn't get scope creeped.

Reply:
Btw, I don't remember off the top of my head and I don't hugely want to go digging; do we have to have things like `Vector() = default` and `Vector & operator=(const Vector &rhs) = default` or can we just get rid of em entirely and cpp is smart enough to figure out that we want defaults for these standard operators?

Reply:
There is no concept of multiple CPU players, there is only one. You're mixing up players with `Teams`.
There can any number of CPU teams, but they will all be controlled by `Player 4`.

In 
```
short ScreenOfPlayer(short player) const { return (player >= Players::PlayerOne && player < Players::MaxPlayerCount) ? m_PlayerScreen[player] : -1; }
```
`player == 4` will never be passed in because 
```
for (short player = Players::PlayerOne; player < Players::MaxPlayerCount; ++player) {
```
The loop will terminate at `player == 3`.
Thus, `(player >= Players::PlayerOne && player < Players::MaxPlayerCount)` is always evaluating to true and `m_PlayerScreen[player]` is returned.
It can only evaluate to false and return -1 if the loop starts from less than 0, terminates after 3, or a bad value is passed manually.
This does not happen.

Now where does it get that -1 from? From the value stored at the array index.

When you start an Activity as P0, they array is {0, -1, -1, -1}. 
P0 is screen index 0 at [0]. P1-3 are disabled with screen index -1 at [1, 2, 3]. *They are not AI players*.
![image](https://user-images.githubusercontent.com/50018358/101223694-49232880-3695-11eb-89f1-f1782206e92b.png)

Here, the first iteration will return 0 from [0], and no crash will happen.
The next iteration will return -1 from [1] and a crash will happen.

When you start the Activity as P1 and P3, the array is {-1, 0, -1, 1}.
P1 is screen index 0 at [1], P3 is screen index 1 at [3]. P0 and P2 are disabled with screen index -1 at [0, 2]. *They are not AI players*.
![image](https://user-images.githubusercontent.com/50018358/101224014-1d547280-3696-11eb-8a44-5287c1d843e6.png)

Here, the first iteration will return -1 from [0] and a crash will happen.

This has absolutely nothing to do with AI teams, or the AI player.
This has to do with `short m_PlayerScreen[Players::MaxPlayerCount];` being what it is.
The AI player is the *fifth player* (P4) and isn't part of anything that deal with human players (P0-P3).

This has everything to do with Dan writing a copy-assignment operator that lets this bullshit slide and because of that, choosing a dumb value for `disabled player` and `disabled player screen index`.

The easiest fix would be changing the disabled screen index from -1 to 5, and increasing the size of `m_PlayerScreen` so it can actually be accessed and assigned to, but never read from because nothing iterates that far.

1. `m_ScreenOcclusion` is also a C array `Vector m_ScreenOcclusion[c_MaxScreenCount];`. There is no `at()`.
Bjarne is telling us to fuck ourselves by letting us access garbage at [-1].

2. Because `&rhs` is coming from a [-1] index it's null and `(*this != rhs)` is true.
![image](https://user-images.githubusercontent.com/50018358/101225004-019e9b80-3699-11eb-8975-f34331a6f8d9.png)

Assignment is happening to nowhere, idfk.
![image](https://user-images.githubusercontent.com/50018358/101225069-2e52b300-3699-11eb-90ed-1fca801bf58a.png)

Reply:
Ok. Interesting stuff. I'll put the old copy-assignment operator back. Hoping that it gets deleted sometime in the future.
I'm assuming that the snippet that this thread is based on should not be removed (until activity.cpp refactor), if I misunderstood let me know.

@garethyr If we don't define the copy-construction operator then we implicitly get the default. 
With the constructor however, we defined one, so if we also want the "default" we have to explicitly say it.